pkg/block/mem/adapter.go

```diff
@@ -24,62 +25,88 @@ var (
 	ErrNoPropertiesForKey = fmt.Errorf("no properties for key")
 )
 
+type partInfo struct {
+	data         []byte
+	lastModified time.Time
+	size         int64
+}
+
 type mpu struct {
-	id    string
-	parts map[int][]byte
+	id        string
+	objectKey string // the key (from getKey(obj))
+	initiated time.Time
+	parts     map[int]partInfo // part number -> partInfo
 }
 
-func newMPU() *mpu {
+func newMPU(objectKey string) *mpu {
 	uid := uuid.New()
 	uploadID := hex.EncodeToString(uid[:])
 	return &mpu{
-		id:    uploadID,
-		parts: make(map[int][]byte),
+		id:        uploadID,
+		objectKey: objectKey,
+		initiated: time.Now(),
+		parts:     make(map[int]partInfo),
 	}
 }
 
 func (m *mpu) get() []byte {
 	buf := bytes.NewBuffer(nil)
-	keys := make([]int, len(m.parts))
-	sort.Slice(keys, func(i, j int) bool {
-		return keys[i] < keys[j]
-	})
+	keys := make([]int, 0, len(m.parts))
+	for k := range m.parts {
+		keys = append(keys, k)
+	}
+	sort.Ints(keys)
 	for _, part := range keys {
-		buf.Write(m.parts[part])
+		buf.Write(m.parts[part].data)
 	}
 	return buf.Bytes()
 }
 
 type Adapter struct {
-	data       map[string][]byte
-	mpu        map[string]*mpu
-	properties map[string]block.Properties
-	mutex      *sync.RWMutex
+	data       map[string]map[string][]byte
+	mpu        map[string]map[string]*mpu
+	properties map[string]map[string]block.Properties
+	mutex      sync.RWMutex
 }
 
 func New(_ context.Context, opts ...func(a *Adapter)) *Adapter {
 	a := &Adapter{
-		data:       make(map[string][]byte),
-		mpu:        make(map[string]*mpu),
-		properties: make(map[string]block.Properties),
-		mutex:      &sync.RWMutex{},
+		data:       make(map[string]map[string][]byte),
+		mpu:        make(map[string]map[string]*mpu),
+		properties: make(map[string]map[string]block.Properties),
 	}
 	for _, opt := range opts {
 		opt(a)
 	}
 	return a
 }
 
+func calcETag(data []byte) string {
+	etag := md5.Sum(data) //nolint:gosec
+	return hex.EncodeToString(etag[:])
+}
+
 func getKey(obj block.ObjectPointer) string {
-	// TODO (niro): Fix mem storage path resolution
-	if obj.IdentifierType == block.IdentifierTypeFull {
+	qk, err := block.DefaultResolveNamespace(obj.StorageNamespace, obj.Identifier, obj.IdentifierType)
```

Comment:
Why this change now?  Does this mean mem didn't work with multiple storage until now?  (Again, obviously not critical, just strange seeing it here.)